## make-DDOL-figures.R by Rohan Maddamsetti.
library(tidyverse)
library(cowplot)


## TODO: make the Supplementary Figures.

## TODO: Notes from Hye-in.
## On the supplementary figures showing the distribution of Nif/Fix genes on plasmids in genomes:
## Add a clear labels for the regions which represent symbiosis plasmids.
## Also, shrink these graphs for the supplement.

#### Line 77: CRITICAL TODO!!!! HANDLE NA VALUES IN metabolic_protein_count!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

## IMPORTANT TODO: make a plot of plasmid distribution in each genome,
## showing all plasmids (gray out the ones without symbiosis genes.)


################################################################################
## Make data structures for the analysis.
## annotate the genomes and plasmids.
gbk.annotation <- read.csv("../results/computationally-annotated-gbk-annotation-table.csv") %>%
    mutate(Annotation = replace_na(Annotation,"Unannotated")) %>%
    ## collapse Annotations into a smaller number of categories as follows:
    ## Marine, Freshwater --> Water
    ## Sediment, Soil, Terrestrial --> Earth
    ## Plants, Agriculture --> Plants
    ## Animals --> Animals
    mutate(Annotation = replace(Annotation, Annotation == "Marine", "Water")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Freshwater", "Water")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Sediment", "Earth")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Soil", "Earth")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Terrestrial", "Earth")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Plants", "Plants")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Agriculture", "Plants")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "Animals", "Animals")) %>%
    mutate(Annotation = replace(Annotation, Annotation == "blank", "Unannotated"))

## get the metadata for chromosomes and plasmids.
replicon.annotation.data <- read.csv("../results/chromosome-plasmid-table.csv") %>%
    rename(SeqType = SequenceType) %>%
    left_join(gbk.annotation) %>%
    ## Annotate the genera.
    mutate(Genus = stringr::word(Organism, 1))

## get mobility data for all plasmids, generated by parse-MOBtyper-results.py.
plasmid.mobility.data <- read.csv("../results/mobility-results.csv")

## Get the length of each plasmid, and number of proteins on them.
plasmid.metadata <- read.csv("../results/replicon-lengths-and-protein-counts.csv") %>%
    filter(SeqType == "plasmid") %>%
    ## rename the Annotation_Accession column for merging with pathway data.
    rename(Annotation_Accession = AnnotationAccession) %>%
    ## replace prefixes.
    mutate(Plasmid = str_replace(SeqID, "N[CZ]_","")) %>%
    ## join the mobility results.
    left_join(plasmid.mobility.data)

## IMPORTANT: This is a tsv file, because "," is a meaningful character in chemical names!   
plasmid.proteins.in.KEGG.metabolism <- read.table("../results/plasmid-proteins-in-KEGG-metabolism.tsv", header = TRUE)

metabolic.genes.per.plasmid <- plasmid.proteins.in.KEGG.metabolism %>%
    group_by(SeqID, SeqType) %>%
    summarize(metabolic_protein_count = n()) %>%
    mutate(metabolic_protein_count = replace_na(metabolic_protein_count, 0))

plasmid.annotation.data <- replicon.annotation.data %>%
    filter(SeqType == "plasmid")

################################################################################
## Figure 6ABC: nitrogen-fixing bacteria are enriched with metabolic genes on plasmids.

#### CRITICAL TODO!!!! HANDLE NA VALUES IN metabolic_protein_count!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

## make the dataframe for Figure 6ABC, showing metabolic genes on plasmids.
metabolic.gene.plot.data <- plasmid.metadata %>%
    left_join(metabolic.genes.per.plasmid) %>%
    ## make the dataframe compatible with plasmid.annotation.data
    mutate(NCBI_Nucleotide_Accession = str_remove(SeqID, "N(C|Z)_")) %>%
    ## and join.
    left_join(plasmid.annotation.data) %>%
    ## remove unannotated bacteria.
    filter(Annotation != "NA") %>%
    filter(Annotation != "Unannotated") %>%
    mutate(Annotation = factor(Annotation)) %>%
    ## Annotate Nitrogen-fixing bacteria.
    mutate(NitrogenFixer = ifelse(
               str_detect(Genus, regex("Rhizo|Azo", ignore_case=TRUE)),
               'Nitrogen-fixer',
               "Non-nitrogen-fixer")) %>%
    arrange(desc(metabolic_protein_count))

## Make the Supplementary File.
write.csv(metabolic.gene.plot.data, "../results/S1File.csv", row.names=FALSE, quote=FALSE)

## Figure 6ABC: plasmids with lots of metabolic proteins come from Plants and Earth.
Fig6A <- metabolic.gene.plot.data %>%
    ggplot(aes(x = metabolic_protein_count, fill = Annotation)) +
    geom_histogram(bins=100) +
    geom_vline(xintercept = 100, color="gray", linetype="dotted") +
    theme_classic() +
    scale_fill_discrete(
        drop=FALSE,
        limits = levels(metabolic.gene.plot.data$Annotation)) +
    xlim(0,500) +
    xlab("") +
    ggtitle("All plasmids with metabolic genes") +
    theme(legend.position="top")

## get the legend and remove from Fig6A.
Fig6ABC_legend <- get_legend(Fig6A)
Fig6A <- Fig6A + guides(fill = "none")

Fig6B <- metabolic.gene.plot.data %>%
    filter(metabolic_protein_count > 100) %>%
    ggplot(aes(x = metabolic_protein_count, fill = Annotation)) +
    geom_histogram(bins=100) +
    geom_vline(xintercept = 100, color="gray", linetype="dotted") +
    theme_classic() +
    scale_fill_discrete(
        drop=FALSE,
        limits = levels(metabolic.gene.plot.data$Annotation)) +
    xlim(0,500) +
    ylim(0,50) +
    xlab("") +
    ggtitle("Plasmids with > 100 metabolic genes") +
    guides(fill = "none")

Fig6C <- metabolic.gene.plot.data %>%
    filter(NitrogenFixer == "Nitrogen-fixer") %>%
    ggplot(aes(x = metabolic_protein_count, fill = Annotation)) +
    geom_histogram(bins=100) +
    geom_vline(xintercept = 100, color="gray", linetype="dotted") +
    theme_classic() +
    scale_fill_discrete(
        drop=FALSE,
        limits = levels(metabolic.gene.plot.data$Annotation)) +
    xlim(0,500) +
    ylim(0,50) +
    xlab("Number of metabolic proteins on the plasmid") +
    ggtitle("Plasmids with metabolic genes in nitrogen-fixing bacteria") +
    guides(fill = "none")

## Save the plot
Fig6ABC <- plot_grid(Fig6A, Fig6B, Fig6C, Fig6ABC_legend,
                      labels=c('A', 'B', 'C', ''), ncol=1, rel_heights=c(1,1,1,0.5))
ggsave("../results/Fig6ABC.pdf", Fig6ABC, width=5.5, height=8)

############################################################
## metabolic genes on plasmids are enriched in nitrogen-fixing bacteria:
## graphical and statistical analysis.

N2.fixer.metabolic.gene.comparison.plot.data <- metabolic.gene.plot.data %>%
    group_by(Annotation_Accession, NitrogenFixer) %>%
    summarize(total_plasmid_metabolic_proteins = sum(metabolic_protein_count))

N2.fixer.comp.data <- N2.fixer.metabolic.gene.comparison.plot.data %>%
    filter(NitrogenFixer=="Nitrogen-fixer")
non.N2.fixer.comp.data <- N2.fixer.metabolic.gene.comparison.plot.data %>%
    filter(NitrogenFixer != "Nitrogen-fixer")

## compare the means of nitrogen-fixing and non-nitrogen-fixing bacteria.
mean.N2.fixer.plasmid.metabolic.genes <- mean(N2.fixer.comp.data$total_plasmid_metabolic_proteins, na.rm=TRUE)
print(mean.N2.fixer.plasmid.metabolic.genes) ## 253.33 plasmid metabolic genes in nitrogen-fixers

mean.non.N2.fixer.plasmid.metabolic.genes <- mean(non.N2.fixer.comp.data$total_plasmid_metabolic_proteins,na.rm=TRUE)
print(mean.non.N2.fixer.plasmid.metabolic.genes) ## 22.97 plasmid metabolic genes in non-nitrogen fixers

## statistics are super significant: p < 1e-93.
wilcox.test(
    x=N2.fixer.comp.data$total_plasmid_metabolic_proteins,
    y=non.N2.fixer.comp.data$total_plasmid_metabolic_proteins,
    alternative="greater")

wilcox.test(
    x=N2.fixer.comp.data$total_plasmid_metabolic_proteins,
    y=non.N2.fixer.comp.data$total_plasmid_metabolic_proteins,
    alternative="greater")$p.value

metabolic.genes.in.N2.fixer.comparison.plot <- N2.fixer.metabolic.gene.comparison.plot.data %>%
    ggplot(aes(x = total_plasmid_metabolic_proteins, y = NitrogenFixer, alpha=0.5)) +
    geom_jitter(width=0,alpha=0.2, size=0.1) +
    geom_boxplot(size=0.1, outlier.shape = NA) +
    theme_classic() +
    guides(alpha = "none") +
    xlab("Numbers of plasmid-borne metabolic genes per genome") +
    ylab("") +
    ggtitle("Plasmids in nitrogen-fixing bacteria are enriched with metabolic genes") +
    theme(plot.title = element_text(size = 10))

## save this boxplot as S1 Supplementary Figure.
ggsave("../results/S1Figure.pdf", metabolic.genes.in.N2.fixer.comparison.plot, width=6, height=1.75)

################################################################################
## Figure 6D: Nif/Fix and Nod genes are co-located on conjugative symbiosis plasmids.

## This analysis and figure examines 454 genomes with Nif/Fix or Nod genes found on plasmids.
## Get the count of genes in each pathway on each plasmid.
NifFix.pathway.count.df <- read.csv("../results/NifFixPlasmidCount.csv")
Nod.pathway.count.df <- read.csv("../results/NodPlasmidCount.csv")
All.symbiosis.pathways.count.df <- read.csv("../results/AllPathwaysPlasmidCount.csv")

## and merge into a big dataframe for analysis.
pathway.plasmid.count.df <- NifFix.pathway.count.df %>%
    full_join(Nod.pathway.count.df) %>%
    full_join(All.symbiosis.pathways.count.df) %>%
    ## add plasmid metadata.
    left_join(plasmid.metadata) %>%
    ## pivot wider to multiply the number of genes in Nif/Fix with the number of genes in Nod.
    ## We define Symbiosis Plasmids as plasmids with BOTH Nif/Fix and Nod genes,
    ## so this multiplication serves as an AND gate.
    pivot_wider(names_from = PathwayType, values_from = PathwayGeneCount) %>%
    ## Now define the SymbiosisPlasmid column.
    mutate(SymbiosisPlasmid = ifelse(`Nif/Fix` * Nod > 0, TRUE, FALSE)) %>%
    ## now pivot longer to get back to the original shape of these data.
    pivot_longer(cols = c("Nif/Fix", "Nod", "All Symbiosis Pathways"),
                 names_to = "PathwayType", values_to = "PathwayGeneCount") %>%
    mutate(PlasmidType = ifelse(SymbiosisPlasmid, "Symbiosis plasmids", "all other plasmids")) %>%
    mutate(PlasmidType = factor(PlasmidType, levels=c( "Symbiosis plasmids", "all other plasmids"))) %>%
    rename(Mobility = PredictedMobility)


## Figure 6D is a pie-chart.
Fig6D.pie.chart.data <- pathway.plasmid.count.df %>%
    group_by(PlasmidType, Mobility) %>%
    summarise(Count = n()) %>%
    group_by(PlasmidType) %>%
    ## Calculate the percentage of each Mobility category
    mutate(Percentage = (Count / sum(Count)) * 100)
## save these data to file.
write.csv(Fig6D.pie.chart.data, "../results/Fig6D-pie-chart-data.csv", row.names=FALSE, quote=FALSE)

## Create the pie chart for Fig 6D.
## Grayson will lay out Figure 6 using Fig6ABC and Fig6D panels by hand.
Fig6D_pie_chart <- Fig6D.pie.chart.data %>%
    ggplot(aes(x = "", y = Percentage, fill = Mobility)) +
    geom_bar(stat = "identity", width = 1) +
    coord_polar(theta = "y") +
    scale_fill_viridis_d(direction = -1) +
    theme_void() +
    facet_grid(PlasmidType~.) +
    theme(legend.position = "bottom") +
    ggtitle("Conjugative symbiosis plasmids\nencode Nif/Fix and Nod pathways\nin nitrogen-fixing bacteria")
## save the pie chart
ggsave("../results/Fig6D-pie-chart.pdf", Fig6D_pie_chart, width=5, height=5)

################################################################################
## The remaining analysis examines 454 genomes with Nif/Fix or Nod genes found on plasmids.
## Symbiosis plasmids with both Nif/Fix and Nod genes are far more likely to be conjugative,
## by comparison with all other plasmids in these 454 genomes.

## 454 genomes here.
print(length(unique(pathway.plasmid.count.df$Annotation_Accession)))

## look at the symbiosis plasmids (Nif/Fix + Nod on these plasmids)
symbiosis.plasmid.data <- pathway.plasmid.count.df %>%
    filter(SymbiosisPlasmid == TRUE)

## 444 Symbiosis plasmids
symbiosis.plasmid.count <- symbiosis.plasmid.data %>%
    nrow() %>%
    print()

## 372 Symbiosis plasmids are conjugative.
conjugative.symbiosis.plasmid.count <- symbiosis.plasmid.data %>%
    filter(Mobility == "conjugative") %>%
    nrow() %>%
    print()

## 24 are mobilizable.
mobilizable.symbiosis.plasmid.count <- symbiosis.plasmid.data %>%
    filter(Mobility == "mobilizable") %>%
    nrow() %>%
    print()

## 48 are non-mobilizable.
non.mobilizable.symbiosis.plasmid.count <- symbiosis.plasmid.data %>%
    filter(Mobility == "non-mobilizable") %>%
    nrow() %>%
    print()

## look at all other plasmids in these 454 genomes.
non.symbiosis.plasmid.data <- pathway.plasmid.count.df %>%
    filter(SymbiosisPlasmid == FALSE)

## 4371 non-symbiosis plasmids in bacteria with Nif/Fix or Nod genes on plasmids.
non.symbiosis.plasmid.count <- non.symbiosis.plasmid.data %>%
    nrow() %>%
    print()

## 1086 non-symbiosis plasmids are conjugative.
conjugative.non.symbiosis.plasmid.count <- non.symbiosis.plasmid.data %>%
    filter(Mobility == "conjugative") %>%
    nrow() %>%
    print()

## 798 are mobilizable.
mobilizable.non.symbiosis.plasmid.count <- non.symbiosis.plasmid.data %>%
    filter(Mobility == "mobilizable") %>%
    nrow() %>%
    print()

## 2487 are non-mobilizable.
non.mobilizable.non.symbiosis.plasmid.count <- non.symbiosis.plasmid.data %>%
    filter(Mobility == "non-mobilizable") %>%
    nrow() %>%
    print()

## so symbiosis plasmids with Nif/Fix and Nod genes are much more likely
## to be conjugative than non-symbiosis plasmids. p-value < 1e-149.
binom.test(x = 372, n = 444, p = (1086/4371))
print(binom.test(x = 372, n = 444, p = (1086/4371))$p.value)

############################################################

## IMPORTANT TODO: make a plot of plasmid distribution in each genome,
## showing all plasmids (gray out the ones without symbiosis genes.)

## count the number of genes found on plasmids in each pathway
## in each genome.
    genome.pathway.gene.count <- pathway.plasmid.count.df %>%
    group_by(Annotation_Accession, PathwayType) %>%
    summarize(TotalPathwayGeneCount = sum(PathwayGeneCount)) %>%
    ungroup()

## rank the genomes by the number of genes found on plasmids
## over all three symbiosis pathways.
genome.all.pathways.ranking <- genome.pathway.gene.count %>%
    filter(PathwayType == "All Symbiosis Pathways") %>%
    ## arrange by total number of symbiosis genes.
    arrange(desc(TotalPathwayGeneCount)) %>%
    ## rank by total number of symbiosis genes.
    mutate(GenomeRank = rank(desc(TotalPathwayGeneCount),ties.method="first")) %>%
    ## drop the PathwayType column for nice merging downstream.
    select(-PathwayType) %>%
    ungroup()

## rank the plasmids containing symbiosis genes in each genome by plasmid length.
symbiosis.plasmid.ranking <- pathway.plasmid.count.df %>%
    filter(PathwayType == "All Symbiosis Pathways") %>%
    ## remove plasmids with no symbiosis genes from the ranking.
    filter(PathwayGeneCount > 0) %>%
    ## rank plasmids containing symbiosis plasmids within each genome based on size.  
    group_by(Annotation_Accession) %>%
    mutate(PlasmidLengthRank = rank(desc(replicon_length),ties.method="first")) %>%
    ## turn PlasmidLengthRank into a factor for plotting.
    mutate(PlasmidLengthRank = as.factor(PlasmidLengthRank)) %>%
    ## for nice merging downstream, we don't want to have the PathwayGeneCount.
    select(Annotation_Accession, Plasmid, replicon_length, protein_count, PlasmidLengthRank) %>%
    ungroup()
  

## add the genome ranks to pathway.plasmid.count.df for plotting.
pathway.plasmid.count.plot.df <- pathway.plasmid.count.df %>%
    ## CRITICAL: remove plasmids with no symbiosis genes from the plot
    ## THIS IS NEEDED FOR CONSISTENCY WITH plasmid.all.pathway.ranking!
    filter(PathwayGeneCount > 0) %>%
    ## add the genome ranking.
    left_join(genome.all.pathways.ranking) %>%
    ## add the plasmid ranking.
    left_join(symbiosis.plasmid.ranking)


stacked.pathway.profile.plot <- pathway.plasmid.count.plot.df %>%
    ## Just show Nif/Fix and Nod.
    filter(PathwayType != "All Symbiosis Pathways") %>%
    ggplot(aes(x=GenomeRank, y = PathwayGeneCount, fill=PlasmidLengthRank)) +
    geom_bar(stat="identity") +
    theme_classic() +
    theme(legend.position="top") +
    facet_grid(PathwayType~.)

stacked.pathway.profile.plot
ggsave("../results/DDOL-pathway-profile.pdf", stacked.pathway.profile.plot, width=8, height = 11)

stacked.pathway.mobility.plot <- pathway.plasmid.count.plot.df %>%
    ## Just show Nif/Fix and Nod.
    filter(PathwayType != "All Symbiosis Pathways") %>%
    ggplot(aes(x=GenomeRank, y = PathwayGeneCount, fill=PredictedMobility)) +
    geom_bar(stat="identity") +
    theme_classic() +
    theme(legend.position="top") +
    facet_grid(PathwayType~.)

stacked.pathway.mobility.plot
ggsave("../results/DDOL-pathway-mobility.pdf", stacked.pathway.mobility.plot, width=8, height = 11)
